### PR TITLE
markdown_preview: Improved markdown rendering support

### PR DESCRIPTION
**Pull request by kierangilliam**
This PR improves support for rendering markdown documents.

## After the updates

https://github.com/zed-industries/zed/assets/18583882/48315901-563d-44c6-8265-8390e8eed942

## Before the updates

https://github.com/zed-industries/zed/assets/18583882/6d7ddb55-41f7-492e-af12-6ab54559f612

## New features

- @SomeoneToIgnore's [scrolling feature request](https://github.com/zed-industries/zed/pull/6958#pullrequestreview-1850458632).
- Checkboxes (`- [ ]` and `- [x]`)
- Inline code blocks.
- Ordered and unordered lists at an arbitrary depth.
- Block quotes that render nested content, like code blocks.
- Lists that render nested content, like code blocks.
- Block quotes that support variable heading sizes and the other markdown features added [here](https://github.com/zed-industries/zed/pull/6958).
- Users can see and click internal links (`[See the docs](./docs.md)`).

## Notable changes

- Removed dependency on `rich_text`.
- Added a new method for parsing markdown into renderable structs. This method uses recursive descent so it can easily support more complex markdown documents.
- Parsing does not happen for every call to `MarkdownPreviewView::render` anymore.

## TODO

- [ ] Typing should move the markdown preview cursor.

## Future work under consideration

- If a title exists for a link, show it on hover.
- Images. 
- Since this PR brings the most support for markdown, we can consolidate `languages/markdown` and `rich_text` to use this new renderer. Note that the updated inline text rendering method in this PR originated from `langauges/markdown`.
- Syntax highlighting in code blocks.
- Footnote references.
- Inline HTML.
- Strikethrough support.
- Scrolling improvements:
    - Handle automatic preview scrolling when multiple cursors are used in the editor.
    - > great to see that the render now respects editor's scrolls, but can we also support the vice-versa (as syntax tree does it in Zed) — when scrolling the render, it would be good to scroll the editor too
    - > sometimes it's hard to understand where the "caret" on the render is, so I wonder if we could go even further with its placement and place it inside the text, as a regular caret? Maybe even support the selections?
- > switching to another markdown tab does not change the rendered contents and when I call the render command again, the screen gets another split — I would rather prefer to have Zed's syntax tree behavior: there's always a single panel that renders things for whatever tab is active now. At least we should not split if there's already a split, rather adding the new rendered tab there.
- > plaintext URLs could get a highlight and the click action

## Release Notes

- Improved support for markdown rendering.

### CODE REVIEW

File: crates/markdown_preview/src/markdown_renderer.rs [hunk 1]
```diff
@@ -1,346 +1,319 @@
-use std::{ops::Range, sync::Arc};
-
+use crate::markdown_elements::{
```
Comment (kierangilliam) [inline crates/markdown_preview/src/markdown_renderer.rs:1]: Most of this file changed so the diff is not helpful. I recommend viewing the file here: https://github.com/kierangilliam/zed/blob/kieran/markdown-parser-and-renderer/crates/markdown_preview/src/markdown_renderer.rs

File: crates/markdown_preview/src/markdown_preview_view.rs [hunk 2]
```diff
@@ -38,29 +44,120 @@ impl MarkdownPreviewView {
...
+                    EditorEvent::Edited => {
+                        let editor = editor.read(cx);
+                        let contents = editor.buffer().read(cx).snapshot(cx).text();
+                        let file_location =
+                            MarkdownPreviewView::get_folder_for_active_editor(editor, cx);
+                        this.contents = parse_markdown(&contents, file_location);
+                        this.list_state.reset(this.contents.children.len());
+                        cx.notify();
+
+                        // TODO: This does not work as expected.
```
Comment (kierangilliam) [inline crates/markdown_preview/src/markdown_preview_view.rs:69]: If anyone has ideas on how to solve this, I'd appreciate the help!

**Comment by kierangilliam** [comment]
> I did do some testing, and one thing I noticed: the text on tasks is gone now.
> 
> ...
>
> I realize tasks weren't implemented before, but the text being gone now is a regression from its previous state.

@JosephTLyons 

For whatever reason, I forgot to include tasks in my playground markdown file. Thanks for the callout. This was an easy add. They are now implemented in this PR:

image

How I render checked / unchecked is not ideal. I can revisit that when I get more free time.

**Review by SomeoneToIgnore — COMMENTED** [review]
Wow, that's a lot of changes!

I have found one notable regression in the way bullet lists are rendered.
Consider https://github.com/rust-lang/rust-analyzer/blob/master/docs/dev/guide.md#ide-api as an example.
Old code is able to show that list, while the new one shows 3 blank dots instead.

Besides that, since it's a huge rewrite, I gave it a spin and here are some items (not necessary to fix now, and some of them are also present in the old version too) that I've found:

* editing the *.md file makes the render to forget its position (I guess that's what the comment near TODO, but I do not know how to help there, sorry)
* it's a subtle thing so feel free to ignore, but I've found adding new carets with alt-click not changing the scroll sometimes (e.g. place the 2nd caret below the current caret — the render goes down, place the 3rd one above all and nothing happens)
* switching to another markdown tab does not change the rendered contents and when I call the render command again, the screen gets another split — I would rather prefer to have Zed's syntax tree behavior: there's always a single panel that renders things for whatever tab is active now. At least we should not split if there's already a split, rather adding the new rendered tab there.
* plaintext URLs could get a highlight and the click action
* great to see that the render now respects editor's scrolls, but can we also support the vice-versa (as syntax tree does it in Zed) — when scrolling the render, it would be good to scroll the editor too
* sometimes it's hard to understand where the "caret" on the render is, so I wonder if we could go even further with its placement and place it inside the text, as a regular caret? Maybe even support the selections?


But apart the missing list rendering, I really like where it goes now: we seem to have more control over the renderer and have tests now!
Thank you for pushing this forward.

**Comment by kierangilliam** [comment]
@SomeoneToIgnore As always — thanks for the thorough feedback. I’ll look at the list rendering issue when I get a chance. 

I agree that the scrolling UX isn’t the best. I’m not sure why the list state doesn’t persist as expected. @mikayla-maki and I had chatted about this previously. Mikayla, do you have any idea how I could work around this?

Would you all prefer deferring scrolling until I can more holistically address your comments?

I agree with the rest of your list. These sound like good follow-up issues.

**Comment by berkus** [comment]
> I would rather prefer to have Zed's syntax tree behavior: there's always a single panel that renders things for whatever tab is active now. At least we should not split if there's already a split, rather adding the new rendered tab there.

Yes! This sounds like a very sane behavior.

File: crates/markdown_preview/src/markdown_preview_view.rs (hunk 2 above)
Comment (berkus) [reply crates/markdown_preview/src/markdown_preview_view.rs:69]: Would it work if cx.notify() is dropped on line 67? (I do not know for sure how notify() works, I just expect it to send whatever updates are pending)

File: crates/markdown_preview/src/markdown_renderer.rs [hunk 3]
```diff
@@ -1,346 +1,322 @@
...
-                    column_alignments,
-                ));
-            }
-            _ => {}
-        }
-    }
+        let bullet = match item.item_type {
+            Ordered(order) => format!("{}.", order),
+            Unordered => "•".to_string(),
+            Task(checked) => if checked { "☑" } else { "☐" }.to_string(),
```
Comment (berkus) [inline crates/markdown_preview/src/markdown_renderer.rs:140]: might be not the best idea, but what about

`Task(checked) => if checked { "✅" } else { "◯" }.to_string(),`

probably too graphical, but pretty!

File: crates/markdown_preview/src/markdown_preview_view.rs (hunk 2 above)
Comment (kierangilliam) [reply crates/markdown_preview/src/markdown_preview_view.rs:69]: Unfortunately not :/

File: crates/markdown_preview/src/markdown_renderer.rs (hunk 3 above)
Comment (SomeoneToIgnore) [reply crates/markdown_preview/src/markdown_renderer.rs:140]: We try to keep Zed from turning into some flashing, rainbow-like thing in its "base", plugin-less version, so not sure if introducing emojis right now is a good thing.

One note on the current style is that it looks somewhat small compared to the rest of my large font (Monocraft)
image

**Review by SomeoneToIgnore — APPROVED** [review]
Nice!
I think we can merge this now.

As for follow-ups, I would propose to concentrate on making the preview to behave more like the syntax tree one:
* change the rendered contents if other *.md tab gets active
* do not create new splits if there's already one
* fix the scroll jump after editing
* scroll the *.md file as the renderer scrolls

But please ignore that and do something else related, if you feel more like it.

**Comment by berkus** [comment]
❤️

**Pull request merged**